### ISSUE TITLE
Question : is it possible to allow deployment when nvidia-smi returns code different from 0

### ISSUE THREAD

**Issue by ymazzer** [NVIDIA/gpu-operator#285]
Hello,

I'm facing some issues trying to make a GPU available in a kubernetes cluster.

Based on my investigations, deployment process stops at nvidia-driver-daemonsets being blocked because the driver-validator which runs `nvidia-smi` will stop the deployment if the command return code is different from 0 but in some case nvidia-smi output can be different from 0 and still be a non blocking warning (for instance return code 14 should not be blocking.)

Is there a way to allow warnings in nvidia-smi output ?

**Comment by kpouget** [comment]
could you show the output of `nvidia-smi; echo $?`, out of curiosity?

**Comment by ymazzer** [comment]
Yes, here it is :

```
| NVIDIA-SMI 470.82.01    Driver Version: 470.82.01    CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla P100-PCIE...  On   | 00000000:0B:00.0 Off |                    0 |
| N/A   42C    P0    29W / 250W |      0MiB / 16280MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
WARNING: infoROM is corrupted at gpu 0000:0B:00.0
14
```

**Comment by kpouget** [comment]
```
RETURN VALUE

       Return code reflects whether the operation succeeded or failed and what
       was the reason of failure.

       Â·      Return code 0 - Success

       Â·      Return code 2 - A supplied argument or flag is invalid
       Â·      Return code 3 - The requested operation is not available on target device
       Â·      Return code 4 - The current user does  not  have permission  to access this device or perform this operation
       Â·      Return code 6 - A query to find an object was unsuccessful
       Â·      Return code  8 - A device's external power cables are not properly attached
       Â·      Return code 9 - NVIDIA driver is not loaded
       Â·      Return code 10 - NVIDIA Kernel detected an interrupt issue  with a GPU
       Â·      Return code 12 - NVML Shared Library couldn't be found or loaded
       Â·      Return code 13 - Local version of NVML  doesn't  implement  this function
       Â·      Return code 14 - infoROM is corrupted
       Â·      Return code 15 - The GPU has fallen off the bus or has otherwise become inaccessible
       Â·      Return code 255 - Other error or internal driver error occurred
```
from https://developer.download.nvidia.com/compute/DCGM/docs/nvidia-smi-367.38.pdf

**Comment by ymazzer** [comment]
@kpouget yes I have read this doc, but the return code 14 is not blocking workloads since it is juste a warning, that is why I'm asking :)

**Comment by kpouget** [comment]
@ymazzer sure, I'm just gathering information around what is happening :)

**Comment by shivamerla** [comment]
@ymazzer Yes, this is interesting issue. Will look into handling these conditions. For now you will need to build(`make ubi8`) a private image of `gpu-operator-validator` from here: https://github.com/NVIDIA/gpu-operator/tree/master/validator after changing `nvidia-smi` references in `main.go` to more specific command (`nvidia-smi -q` or `nvidia-smi -L`) which is not throwing this warning.  If nothing works you can use `nvidia-smi -h` to workaround meanwhile we provide a way to handle this.

**Comment by zvonkok** [comment]
@ymazzer Is you have a running driver-container, look for errors/warnings in the device-plugin, this might fail as well if `nvidia-smi` is providing a non-zero exit code.

**Comment by ymazzer** [comment]
Thanks for your support, I managed to make it work with a private build.

Do you plan to make it working natively? should I make a pull request?

@zvonkok I didn't use the device plugin since the driver and the toolkit are installed on the host.

**Comment by shivamerla** [comment]
@kpouget created PR for this: https://gitlab.com/nvidia/kubernetes/gpu-operator/-/merge_requests/354

Not likely that it will get into v1.9.0 release, will update.

**Comment by shivamerla** [comment]
@kpouget can you try out with this validator image `registry.gitlab.com/shivamerla/gpu-operator/gpu-operator-validator:14ab83e6c2901d94736a52c12f82c9ce0b21a569-ubi8` and passing ENV as done in the PR. We need to make sure plugin/GFD can handle these errors.

**Comment by kpouget** [comment]
@shivamerla, the image you link is the build of the [PR](https://gitlab.com/nvidia/kubernetes/gpu-operator/-/merge_requests/354), right?
I'll give it a try tomorrow in the CI (we build the validator image there)

**Comment by shivamerla** [comment]
yes, thats correct its from the CI.

**Comment by kpouget** [comment]
@shivamerla, couldn't the driver Pod check for `nvidia-smi` to return `0` before going for its infinite sleep?

currently, on a broken GPU, the [driver](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/periodic-ci-rh-ecosystem-edge-ci-artifacts-master-4.10-nvidia-gpu-operator-e2e-1-7-x/1495533928654573568/artifacts/nvidia-gpu-operator-e2e-1-7-x/nightly/artifacts/013__gpu_operator__capture_deployment_state/gpu_operator_nvidia-driver-daemonset.log) Pod is happy, not mentioning any issue:
```
[pod/nvidia-driver-daemonset-8l7bp/nvidia-driver-ctr] Done, now waiting for signal
[pod/nvidia-driver-daemonset-8l7bp/nvidia-driver-ctr] + trap 'echo '\''Caught signal'\''; _shutdown && { kill 17979; exit 0; }' HUP INT QUIT PIPE TERM
[pod/nvidia-driver-daemonset-8l7bp/nvidia-driver-ctr] + sleep infinity
[pod/nvidia-driver-daemonset-8l7bp/nvidia-driver-ctr] + trap - EXIT
[pod/nvidia-driver-daemonset-8l7bp/nvidia-driver-ctr] + true
[pod/nvidia-driver-daemonset-8l7bp/nvidia-driver-ctr] + wait 17979
```

but the validation [test](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/periodic-ci-rh-ecosystem-edge-ci-artifacts-master-4.10-nvidia-gpu-operator-e2e-1-7-x/1495533928654573568/artifacts/nvidia-gpu-operator-e2e-1-7-x/gather-extra/artifacts/pods/gpu-operator-resources_nvidia-container-toolkit-daemonset-gj6sg_driver-validation.log) fails:
```
WARNING: infoROM is corrupted at gpu 0000:00:1E.0
command failed, retrying after 5 seconds
```

and this log isn't even captured in by the must-gather script, 
I'm not exactly sure why but the validation pod log ends up [empty](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/periodic-ci-rh-ecosystem-edge-ci-artifacts-master-4.10-nvidia-gpu-operator-e2e-1-7-x/1495533928654573568/artifacts/nvidia-gpu-operator-e2e-1-7-x/nightly/artifacts/013__gpu_operator__capture_deployment_state/gpu_operator_nvidia-operator-validator.log) (certainly because the validation is done in an init pod, but I'm not sure what it's not captured by the `-all-containers --prefix` flags)

I think it could be simple and efficient to block the driver Pod from reporting as ready, and have the  `WARNING: infoROM is corrupted at gpu 0000:00:1E.0` in the logs

**Comment by cdesiniotis** [comment]
The last update on this issue was over 3 years ago and the context may now be outdated.

In an effort to clean up our issue tracker and focus on current, actionable issues, I am going to close this. But feel free to re-open or file a new ticket with updated information if this issue is still relevant with latest versions of GPU Operator.
